pdg/src/main.rs

```diff
@@ -24,34 +24,43 @@ mod query;
 mod util;
 
 use builder::{construct_pdg, read_event_log};
-use c2rust_analysis_rt::Runtime;
 use color_eyre::eyre;
 use std::{env, path::Path};
 
+use crate::{builder::read_metadata, graph::IWithMetadata};
+
 fn main() -> eyre::Result<()> {
     color_eyre::install()?;
     env_logger::init();
-    let _runtime = Runtime::new();
 
-    let event_trace_path = env::args()
+    let metadata_path = env::args_os()
+        .nth(2)
+        .expect("Expected metadata file path as the 1st argument");
+    let event_trace_path = env::args_os()
         .nth(1)
-        .expect("Expected event trace file path as the first argument");
-    let events = read_event_log(Path::new(event_trace_path.as_str()))?;
+        .expect("Expected event trace file path as the 2nd argument");
+
+    let metadata = read_metadata(Path::new(&metadata_path))?;
+    let events = read_event_log(Path::new(&event_trace_path))?;
 
     // for event in &events {
-    //     let mir_loc = mir_loc::get(event.mir_loc).unwrap();
+    //     use c2rust_analysis_rt::metadata::IWithMetadata;
+
+    //     let mir_loc = metadata.get(event.mir_loc);
+    //     let mir_loc = mir_loc.with_metadata(&metadata);
     //     let kind = &event.kind;
     //     println!("{mir_loc:?} -> {kind:?}");
     // }
 
-    let pdg = construct_pdg(&events);
+    let pdg = construct_pdg(&events, &metadata);
     pdg.assert_all_tests();
 
     for (graph_id, graph) in pdg.graphs.iter_enumerated() {
         let needs_write = graph
             .needs_write_permission()
             .map(|node_id| node_id.as_usize())
             .collect::<Vec<_>>();
+        let graph = graph.with_metadata(&metadata);
```

Comment:
That is true I think.  If we replace

```rust
pub struct Func(pub DefPathHash);
```

with

```rust
pub struct Func {
    pub def_path_hash: DefPathHash,
    pub name: String,
}
```

which we create with the `Metadata` while constructing the PDG, then I think we'll be able to get rid of the `WithMetadata` stuff for the PDG code.  We still need it in some places, but not directly in `c2rust-pdg`.

Should I make that change?  Thanks for pointing this out!

Reply:
Fixed in https://github.com/immunant/c2rust/pull/490/commits/eff8ff9a47459803f64c3157629602d38c67a934.  Does that look better?  I think it's a lot cleaner now.

Reply:
It's a great start; it could be taken a step further where you add the function name to the `MirLoc` during instrumentation, or just before it ends, and use that in creating `Node.function` -- that would preclude the need for `WithMetadata` etc. altogether.

Reply:
Ultimately I do think we still need it because we don't want to put extra data in `Event` and we want to have a `Debug` `impl` for it, which needs the `Metadata`.

Reply:
That's basically the same logic though.  I don't think getting rid of WithMetadata is worth that much, as then we also lose the debugging convenience of `Debug` `impl`s.

Reply:
Something like #496 is what I'd be willing to accept. It removes code and has the same logic.

Reply:
https://github.com/immunant/c2rust/pull/496 is merged now.